activation/post_service.go

```diff
@@ -0,0 +1,148 @@
+package activation
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os/exec"
+	"path/filepath"
+	"sync"
+
+	"github.com/spacemeshos/post/config"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/events"
+)
+
+// DefaultPostServiceConfig returns the default config for post service. These are intended for testing.
+func DefaultPostServiceConfig() PostServiceConfig {
+	cfg := PostServiceConfig{
+		PostServiceCmd:  "/bin/service",
+		DataDir:         config.DefaultDataDir,
+		NodeAddress:     "http://127.0.0.1:9093",
+		PowDifficulty:   config.DefaultConfig().PowDifficulty,
+		PostServiceMode: "light",
+	}
+
+	return cfg
+}
+
+// MainnetPostServiceConfig returns the default config for mainnet.
+func MainnetPostServiceConfig() PostServiceConfig {
+	cfg := PostServiceConfig{
+		PostServiceCmd:  "/bin/service",
+		DataDir:         config.DefaultDataDir,
+		NodeAddress:     "http://127.0.0.1:9093",
+		PowDifficulty:   config.MainnetConfig().PowDifficulty,
+		PostServiceMode: "fast",
+	}
+
+	return cfg
+}
+
+type PostServiceConfig struct {
+	PostServiceCmd string `mapstructure:"post-opts-post-service"`
+
+	DataDir         string        `mapstructure:"post-opts-datadir"`
+	NodeAddress     string        `mapstructure:"post-opts-node-address"`
+	PowDifficulty   PowDifficulty `mapstructure:"post-opts-pow-difficulty"`
+	PostServiceMode string        `mapstructure:"post-opts-post-service-mode"`
+}
+
+// PostService manages a local post service.
+type PostService struct {
+	logger *zap.Logger
+
+	pidMtx sync.RWMutex
+	pid    int // pid of the running post service.
+
+	stop context.CancelFunc // stops the running command.
+	eg   errgroup.Group     // eg manages post service goroutines.
+}
+
+// NewPostService returns a new post service.
+func NewPostService(logger *zap.Logger, opts PostServiceConfig) (*PostService, error) {
+	ctx, stop := context.WithCancel(context.Background())
+
+	ps := &PostService{
+		logger: logger,
+		stop:   stop,
+	}
+	ps.eg.Go(func() error { return ps.runCmd(ctx, opts) })
+	return ps, nil
+}
+
+func (ps *PostService) Close() error {
+	ps.stop()
+	ps.eg.Wait()
+	return nil
+}
+
+// captureCmdOutput returns a function that reads from the given pipe and logs the output.
+// it returns when the pipe is closed.
+func (ps *PostService) captureCmdOutput(pipe io.ReadCloser) func() error {
+	return func() error {
+		for {
+			buf := make([]byte, 1024)
+			n, err := pipe.Read(buf)
+			switch err {
```

Comment:
Would it make sense to reuse the `buf`?
```suggestion
		buf := make([]byte, 1024)
		for {
			n, err := pipe.Read(buf)
			switch err {
```

Reply:
I ended up using a `bufio.Reader` instead of manually processing the output in chunks.

This also solved a cosmetic issue I had with the logs where a lot of empty lines where printed.